File: src/main/java/com/group1/stagesWs/service/ClasseService.java
package com.group1.stagesWs.service;

import com.group1.stagesWs.model.*;
import com.group1.stagesWs.repositories.*;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Random;

@Service
public class ClasseService {

    private final AllerRetourRepository allerRetourRepository;
    private final BarilRepository barilRepository;
    private final TourRepository tourRepository;
    private final CompetitionRepository competitionRepository;
    private final ChevalRepository chevalRepository;
    private final EquipeRepository equipeRepository;
    private final ParticipantRepository participantRepository;

    public ClasseService(AllerRetourRepository allerRetourRepository, BarilRepository barilRepository, TourRepository tourRepository, CompetitionRepository competitionRepository, ChevalRepository chevalRepository, EquipeRepository equipeRepository, ParticipantRepository participantRepository) {
        this.allerRetourRepository = allerRetourRepository;
        this.barilRepository = barilRepository;
        this.tourRepository = tourRepository;
        this.competitionRepository = competitionRepository;
        this.chevalRepository = chevalRepository;
        this.equipeRepository = equipeRepository;
        this.participantRepository = participantRepository;
    }

    public Optional<AllerRetour> createAllerRetourClasse(AllerRetour classe, int competitionId) {
        classe.setNom("Aller retour");
        AllerRetour finalClasse = allerRetourRepository.save(classe);
        Competition competition = competitionRepository.getById(competitionId);
        competition.setAllerRetour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Baril> createBarilClasse(Baril classe, int competitionId) {
        classe.setNom("Course de Barils");
        Baril finalClasse = barilRepository.save(classe);
        Competition competition = competitionRepository.getById(competitionId);
        competition.setBaril(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Tour> createTourClasse(Tour classe, int competitionId) {
        classe.setNom("Tour de ring");
        Tour finalClasse = tourRepository.save(classe);
        Competition competition = competitionRepository.getById(competitionId);
        competition.setTour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    private Equipe generateEquipe(int chevalId, String courriel) {
        Cheval cheval = chevalRepository.findById(chevalId);
        Participant participant = participantRepository.findParticipantByCourrielIgnoreCase(courriel);
        return new Equipe(participant, cheval);
    }

    private List<Equipe> generateInscriptionList(Classe classe, Equipe equipe) {
        List<Equipe> equipeList = classe.getInscriptionList();
        if (equipeList.isEmpty()) {
            equipeList = new ArrayList<>(List.of(equipe));
        } else {
            equipeList.add(equipe);
        }
        return equipeList;
    }

    public Optional<AllerRetour> sinscrireAllerRetour(int chevalId, String courriel, int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        Competition competition = competitionRepository.findByAllerRetour(classe);
        Equipe equipe = generateEquipe(chevalId, courriel);
        equipeRepository.save(generateEquipe(chevalId, courriel));
        classe.setInscriptionList(generateInscriptionList(classe, equipe));
        AllerRetour finalClasse = allerRetourRepository.save(classe);
        competition.setAllerRetour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Baril> sinscrireBaril(int chevalId, String courriel, int classeId) {
        Baril classe = barilRepository.findById(classeId);
        Competition competition = competitionRepository.findByBaril(classe);
        Equipe equipe = generateEquipe(chevalId, courriel);
        equipeRepository.save(equipe);
        classe.setInscriptionList(generateInscriptionList(classe, equipe));
        Baril finalClasse = barilRepository.save(classe);
        competition.setBaril(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Tour> sinscrireTour(int chevalId, String courriel, int classeId) {
        Tour classe = tourRepository.findById(classeId);
        Competition competition = competitionRepository.findByTour(classe);
        Equipe equipe = generateEquipe(chevalId, courriel);
        equipeRepository.save(equipe);
        classe.setInscriptionList(generateInscriptionList(classe, equipe));
        Tour finalClasse = tourRepository.save(classe);
        competition.setTour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<AllerRetour> genererOrdreAllerRetour(int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        Competition competition = competitionRepository.findByAllerRetour(classe);
        List<Equipe> equipeList = classe.getInscriptionList();
        List<Equipe> ordre = genererOrdre(equipeList);
        classe.setOrdreDePassage(ordre);
        AllerRetour finalClasse = allerRetourRepository.save(classe);
        competition.setAllerRetour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Baril> genererOrdreBaril(int classeId) {
        Baril classe = barilRepository.findById(classeId);
        Competition competition = competitionRepository.findByBaril(classe);
        List<Equipe> equipeList = classe.getInscriptionList();
        List<Equipe> ordre = genererOrdre(equipeList);
        classe.setOrdreDePassage(ordre);
        Baril finalClasse = barilRepository.save(classe);
        competition.setBaril(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public Optional<Tour> genererOrdreTour(int classeId) {
        Tour classe = tourRepository.findById(classeId);
        Competition competition = competitionRepository.findByTour(classe);
        List<Equipe> equipeList = classe.getInscriptionList();
        List<Equipe> ordre = genererOrdre(equipeList);
        classe.setOrdreDePassage(ordre);
        Tour finalClasse = tourRepository.save(classe);
        competition.setTour(finalClasse);
        competitionRepository.save(competition);
        return Optional.of(finalClasse);
    }

    public List<Equipe> genererOrdre(List<Equipe> equipeList) {
        Random random_method = new Random();
        List<Equipe> ordre = new ArrayList<>();
        for (int i = 0; i < equipeList.size(); i++) {
            int index = random_method.nextInt(equipeList.size());
            ordre.add(equipeList.get(index));
        }
        return ordre;
    }

    public List<Equipe> getOrdreAllerRetour(int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        return classe.getOrdreDePassage();
    }


    public List<Equipe> getResultatAllerRetour(int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        return classe.getClassement();
    }

    public Integer getEquipeActuelIdAllerRetour(int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        List<Equipe> resultatList = classe.getClassement();
        return resultatList.size();
    }

    public Integer getEquipeActuelIdBaril(int classeId) {
        Baril classe = barilRepository.findById(classeId);
        List<Equipe> resultatList = classe.getClassement();
        return resultatList.size();
    }

    public Integer getEquipeActuelIdTour(int classeId) {
        Tour classe = tourRepository.findById(classeId);
        List<Equipe> resultatList = classe.getClassement();
        return resultatList.size();
    }

    public List<Cheval> getOrdreListChevalTour(int classeId, int equipeId) {
        Tour classe = tourRepository.findById(classeId);
        List<Cheval> tempListOrdreCheval = new ArrayList<>();
        for (int i = equipeId + 1; i < classe.getOrdreDePassage().size(); i++) {
            tempListOrdreCheval.add(classe.getOrdreDePassage().get(i).getCheval());
        }
        return tempListOrdreCheval;
    }

    public List<EquipeAfficher> getResultatListAllerRetour(int classeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        return getResultatList(classe);
    }

    public List<EquipeAfficher> getResultatListBaril(int classeId) {
        Baril classe = barilRepository.findById(classeId);
        return getResultatList(classe);
    }

    public List<EquipeAfficher> getResultatListTour(int classeId) {
        Tour classe = tourRepository.findById(classeId);
        return getResultatList(classe);
    }

    private List<EquipeAfficher> getResultatList(Classe classe) {
        List<EquipeAfficher> tempListResultat = new ArrayList<>();
        for (Equipe equipe : classe.getClassement()) {
            tempListResultat.add(new EquipeAfficher(equipe.getId(), equipe.getParticipant().getPrenom(), equipe.getParticipant().getNom(), equipe.getCheval().getNom(), equipe.getTemps()));
        }
        return tempListResultat;
    }

    public List<EquipeAfficher> getOrdreListAllerRetour(int classeId, int equipeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        return getOrdreList(classe, equipeId);
    }

    public List<EquipeAfficher> getOrdreListBaril(int classeId, int equipeId) {
        Baril classe = barilRepository.findById(classeId);
        return getOrdreList(classe, equipeId);
    }

    public List<EquipeAfficher> getOrdreListTour(int classeId, int equipeId) {
        Tour classe = tourRepository.findById(classeId);
        return getOrdreList(classe, equipeId);
    }

    private List<EquipeAfficher> getOrdreList(Classe classe, int equipeId) {
        List<EquipeAfficher> tempListOrdre = new ArrayList<>();
        for (int i = equipeId + 1; i < classe.getOrdreDePassage().size(); i++) {
            tempListOrdre.add(new EquipeAfficher(classe.getOrdreDePassage().get(i).getId(), classe.getOrdreDePassage().get(i).getParticipant().getPrenom(), classe.getOrdreDePassage().get(i).getParticipant().getNom(), classe.getOrdreDePassage().get(i).getCheval().getNom(), classe.getOrdreDePassage().get(i).getTemps()));
        }
        return tempListOrdre;
    }

    public EquipeAfficher getEquipeActuelAllerRetour(int classeId, int equipeId) {
        AllerRetour classe = allerRetourRepository.findById(classeId);
        return getEquipeActuel(classe, equipeId);
    }

    public EquipeAfficher getEquipeActuelBaril(int classeId, int equipeId) {
        Baril classe = barilRepository.findById(classeId);
        return getEquipeActuel(classe, equipeId);
    }

    public EquipeAfficher getEquipeActuelTour(int classeId, int equipeId) {
        Tour classe = tourRepository.findById(classeId);
        return getEquipeActuel(classe, equipeId);
    }

    private EquipeAfficher getEquipeActuel(Classe classe, int equipeId) {
        Equipe equipe = classe.getOrdreDePassage().get(equipeId);
        return new EquipeAfficher(equipe.getId(), equipe.getParticipant().getPrenom(), equipe.getParticipant().getNom(), equipe.getCheval().getNom(), equipe.getTemps());
    }
}

File: react-stage/src/services/ClasseService.js
const urlBase = 'http://localhost:9191/classe'

const ClasseService = {
    saveAllerRetour: async (values, competitionId) => {
        const res = await fetch(urlBase + '/AllerRetour/' + competitionId,
            {
                method: 'POST',
                headers: {
                    'Content-type': 'application/json'
                },
                body: JSON.stringify(values)
            }
        )
        const data = await res.json()
        return data
    },

    saveTour: async (values, competitionId) => {
        const res = await fetch(urlBase + '/Tour/' + competitionId,
            {
                method: 'POST',
                headers: {
                    'Content-type': 'application/json'
                },
                body: JSON.stringify(values)
            }
        )
        const data = await res.json()
        return data
    },

    saveBaril: async (values, competitionId) => {
        const res = await fetch(urlBase + '/Baril/' + competitionId,
            {
                method: 'POST',
                headers: {
                    'Content-type': 'application/json'
                },
                body: JSON.stringify(values)
            }
        )
        const data = await res.json()
        return data
    },

    sinscrireAllerRetour: async (chevalId, courriel, classeId) => {
        const res = await fetch(urlBase + "/inscrireAllerRetour/" + chevalId + "/" + courriel + "/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    sinscrireBaril: async (chevalId, courriel, classeId) => {
        const res = await fetch(urlBase + "/inscrireBaril/" + chevalId + "/" + courriel + "/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    sinscrireTour: async (chevalId, courriel, classeId) => {
        const res = await fetch(urlBase + "/inscrireTour/" + chevalId + "/" + courriel + "/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    genererOrdreAllerRetour: async (classeId) => {
        const res = await fetch(urlBase + "/genererOrdreAllerRetour/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    genererOrdreBaril: async (classeId) => {
        const res = await fetch(urlBase + "/genererOrdreBaril/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    genererOrdreTour: async (classeId) => {
        const res = await fetch(urlBase + "/genererOrdreTour/" + classeId,
            {
                method: 'POST'
            })
        const data = await res.json()
        return data
    },

    getEquipeActuelIdAllerRetour: async (classeId) => {
        const res = await fetch(urlBase + '/getEquipeActuelIdAllerRetour/' + classeId)
        const data = await res.json()
        return data
    },

    getEquipeActuelIdBaril: async (classeId) => {
        const res = await fetch(urlBase + '/getEquipeActuelIdBaril/' + classeId)
        const data = await res.json()
        return data
    },

    getEquipeActuelIdTour: async (classeId) => {
        const res = await fetch(urlBase + '/getEquipeActuelIdTour/' + classeId)
        const data = await res.json()
        return data
    },

    getResultatListAllerRetour: async (classeId) => {
        const res = await fetch(urlBase + '/getResultatListAllerRetour/' + classeId)
        const data = await res.json()
        return data
    },

    getResultatListBaril: async (classeId) => {
        const res = await fetch(urlBase + '/getResultatListBaril/' + classeId)
        const data = await res.json()
        return data
    },

    getResultatListTour: async (classeId) => {
        const res = await fetch(urlBase + '/getResultatListTour/' + classeId)
        const data = await res.json()
        return data
    },

    getOrdreListAllerRetour: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getOrdreListAllerRetour/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },

    getOrdreListBaril: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getOrdreListBaril/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },

    getOrdreListTour: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getOrdreListTour/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },

    getEquipeActuelAllerRetour: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getEquipeActuelAllerRetour/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },

    getEquipeActuelBaril: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getEquipeActuelBaril/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },

    getEquipeActuelTour: async (classeId, equipeActuelId) => {
        const res = await fetch(urlBase + '/getEquipeActuelTour/' + classeId + '/' + equipeActuelId)
        const data = await res.json()
        return data
    },
}
export default ClasseService